zilliqa/src/exec.rs

```diff
@@ -180,20 +186,25 @@ impl State {
         let mut created_contract_addr: Option<Address> = None;
 
         let mut code: Vec<u8> = account.code;
-        let mut data: Vec<u8> = payload;
+        let mut data: Vec<u8> = payload.clone();
         let mut traces: Arc<Mutex<LoggingEventListener>> =
             Arc::new(Mutex::new(LoggingEventListener::new(tracing)));
 
         // The backend is provided to the evm as a way to read accounts and state during execution
         let mut backend = EvmBackend::new(self, U256::zero(), caller, chain_id, current_block);
 
         // if this is none, it is contract creation
-        if to_addr.is_none() {
+        if to_addr.is_none() || to_addr.is_some_and(|x| x == Address::zero()) {
```

Comment:
This isn't correct. In Ethereum-land you are allowed to send a transaction to the zero address and it means something different to a transaction without an address.

Reply:
The correct way to create a contract in eth is to send a zero length byte array for the 'to'. Thus it should not be Address::zero(), that is correct. However, I believe that certain clients will send the zero address to indicate contract creation (including scilla). Would you agree?

This isn't right though as you point out, because someone might be trying to burn tokens, and there is not neccessarily data with the tx.

Do you think that when the scilla TX is received, we should map the zero address to Option<None>?

Reply:
> However, I believe that certain clients will send the zero address to indicate contract creation (including scilla). Would you agree?

I think its important to consider the Ethereum and Zilliqa APIs separately.

In Ethereum, `to_addr = None` and `to_addr = Some(0x0)` are two different things - The transaction payloads will be different, have different signatures and different semantics. Clients that confuse these will definitely break on other EVM chains.

In Zilliqa, `to_addr` is not `Option`al and contract creations are specified by setting `to_addr = 0x0`.

I think the correct place for this mapping would be in the `fn to_addr` method of `SignedTransaction`. I think we should change the `Zilliqa` variant from:
```rust
            Transaction::Zilliqa(TxZilliqa { to_addr, .. }) => Some(*to_addr),
```
to
```rust
            Transaction::Zilliqa(TxZilliqa { to_addr, .. }) if to_addr.is_zero() => None,
            Transaction::Zilliqa(TxZilliqa { to_addr, .. }) => Some(*to_addr),
```